diff --git a/src/Tensor.cpp b/src/Tensor.cpp
index 600267c..0e8a4f0 100644
--- a/src/Tensor.cpp
+++ b/src/Tensor.cpp
@@ -275,14 +275,16 @@ namespace kp {
     case MemoryTypes::eDeviceAndHost:
       return vk::BufferUsageFlagBits::eStorageBuffer |
         vk::BufferUsageFlagBits::eTransferSrc |
-        vk::BufferUsageFlagBits::eTransferDst;
+        vk::BufferUsageFlagBits::eTransferDst |
+        vk::BufferUsageFlagBits::eIndirectBuffer;
       break;
     case MemoryTypes::eStorage:
       return vk::BufferUsageFlagBits::eStorageBuffer |
         // You can still copy buffers to/from storage memory
         // so set the transfer usage flags here.
         vk::BufferUsageFlagBits::eTransferSrc |
-        vk::BufferUsageFlagBits::eTransferDst;
+        vk::BufferUsageFlagBits::eTransferDst |
+        vk::BufferUsageFlagBits::eIndirectBuffer;
       break;
     default:
       throw std::runtime_error("Kompute Tensor invalid tensor type");
