diff --git a/src/include/kompute/Manager.hpp b/src/include/kompute/Manager.hpp
index a73a45b..20d7147 100644
--- a/src/include/kompute/Manager.hpp
+++ b/src/include/kompute/Manager.hpp
@@ -510,6 +510,17 @@ namespace kp {
      **/
     std::shared_ptr<vk::Instance> getVkInstance() const;
 
+    /**
+     * Get the logical device
+     * */
+    std::shared_ptr<vk::Device> getDevice() {
+      return this->mDevice;
+    }
+
+    std::shared_ptr<vk::PhysicalDevice> getPhysicalDevice() {
+      return this->mPhysicalDevice;
+    }
+
   private:
     // -------------- OPTIONALLY OWNED RESOURCES
     std::shared_ptr<vk::Instance> mInstance = nullptr;
